enhancement(stdlib): Add optional `http_proxy` and `https_proxy` params to `http_request`

5Dev24 opened the pull request:
## Summary

Adds optional parameters for setting a proxy for http and https requests to `http_request`.

A future enhancement should use the [global proxy](https://github.com/vectordotdev/vector/blob/cd2471ab3ed81cc55b13eb4f094af735a210b61e/lib/vector-core/src/config/global_options.rs#L103) by default if set. Is there any easy way to pass that information from vector to vrl?

## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

Locally testing since I didn't want to add standing up a whole proxy server as party of the test suite.

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on
  our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [ ] Our [CONTRIBUTING.md](https://github.com/vectordotdev/vrl/blob/main/CONTRIBUTING.md) is a good starting place.
- [ ] If this PR introduces changes to [LICENSE-3rdparty.csv](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv), please
  run `dd-rust-license-tool write` and commit the changes. More details [here](https://crates.io/crates/dd-rust-license-tool).
- [ ] For new VRL functions, please also create a sibling PR in Vector to document the new function.

## References

None

5Dev24 committed:
Add optional proxy params to http_request

5Dev24 committed:
Add changelog fragment

pront commented on src/stdlib/http_request.rs:
```diff
@@ -73,37 +89,89 @@ mod non_wasm {
+
+        if let Some(http_proxy) = http_proxy {
+            proxies.push(
+                Proxy::http(http_proxy.try_bytes_utf8_lossy()?.into_owned())
```
`into_owned` looks redundant here

pront commented on src/stdlib/http_request.rs:
```diff
@@ -73,37 +89,89 @@ mod non_wasm {
+                .transpose()?;
+
+            let client = if let Some(proxies) = make_proxies(http_proxy, https_proxy)? {
+                &build_client(Some(proxies))
```
We are building a client on each request here. It should be possible to detect literals in `compile`, call `resolve_constant` and create the client there.

pront reviewed (commented):
A few problems but overall looks good

5Dev24 committed:
Address review issues

Pre-existing: remove redundant `to_uppercase`
Avoid `into_owned` cloning of proxy values
Attempt to build http client when proxies are constants

pront reviewed (commented):
forgot to submit these comments

5Dev24 committed:
Explicitly cast error result

pront committed:
remove RefOrOwnedClient

pront reviewed (approved):
Pushed commit 7a5cdad86cf1a72202743a5ba6662275d6bfbe25.
@5Dev24 this looks good to merge now. 
Feel free to follow up if something was missed.

pront committed:
fmt
